In [2]:
from subgrounds.subgrounds import Subgrounds
from subgrounds.subgraph import SyntheticField
from datetime import datetime
from subgrounds.plotly_wrappers import Bar, Figure
from subgrounds.dash_wrappers import Graph

import pandas as pd

# never used subgrounds so getting used to the syntax here

In [3]:
# initialize Subgrounds object
sg = Subgrounds()

# load livepeer subgraph
lpt = sg.load_subgraph('https://api.thegraph.com/subgraphs/name/livepeer/livepeer')

In [4]:
# define FieldPaths
transcoders = lpt.Query.transcoders
delegators = lpt.Query.delegators


In [17]:
top_10_transcoders = transcoders(
    orderBy=transcoders.totalStake,
    orderDirection='desc',
    first=10,
    where=[
        transcoders.active == True
    ]
)

sg.query_df([
    top_10_transcoders.id,
    top_10_transcoders.totalStake,
    top_10_transcoders.totalVolumeETH,
    top_10_transcoders.pricePerSegment
])

transcoders_id  transcoders_totalStake  \
0  0x9c10672cee058fd658103d90872fe431bb6c0afa            2.516184e+06   
1  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9            1.252157e+06   
2  0x525419ff5707190389bfb5c87c375d710f5fcb0e            9.233017e+05   
3  0x4ff088ac5422f994486663ff903b040692797168            7.841820e+05   
4  0x942f0c28fb85ea0b50bfb76a3ecfa99861fa9b4b            7.284690e+05   
5  0xe0a4a877cd0a07da7c08dffebc2546a4713147f2            6.424943e+05   
6  0x4f4758f7167b18e1f5b3c1a7575e3eb584894dbc            5.444993e+05   
7  0x731808ad8b1c3d13e8972db838ada5fc6ae3c2c8            5.117972e+05   
8  0xd84781e1a9b74d71ea76cda8bb9f30893bfd00d1            4.195702e+05   
9  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd            3.410711e+05   

   transcoders_totalVolumeETH  transcoders_pricePerSegment  
0                    8.001220                 150000000000  
1                    0.722895                 200000000000  
2                   21.328402                 150000000000  
3                    0.650195                 200000000000  
4                    3.114552                            0  
5                    0.000000                 150000000000  
6                    0.628243                 171000000000  
7                    0.000000                            0  
8                    1.093463                           10  
9                   10.829665                    123456789

In [59]:
# call ratio
active_transcoders = transcoders(
    orderBy=transcoders.totalStake,
    orderDirection='desc',
    first=10,
    where=[
        transcoders.active == True
    ]
)

# subgrounds does not really account for null as an option
# solution is to grab the last 30 non-null and max-min will tell us how many were null
active_pools = active_transcoders.pools(
    first=30,
    where=[
        active_transcoders.pools.rewardTokens != None
    ]
)

pool_rewards = sg.query_df([
    active_transcoders.id,
    active_pools.round.id,
    active_pools.rewardTokens
])

pool_rewards

# req = sg.mk_request([
#     active_transcoders.id,
#     active_pools.round.id,
#     active_pools.rewardTokens
# ])

# print(req.graphql)

transcoders_id transcoders_pools_round_id  \
0    0x9c10672cee058fd658103d90872fe431bb6c0afa                       1317   
1    0x9c10672cee058fd658103d90872fe431bb6c0afa                       1318   
2    0x9c10672cee058fd658103d90872fe431bb6c0afa                       1319   
3    0x9c10672cee058fd658103d90872fe431bb6c0afa                       1320   
4    0x9c10672cee058fd658103d90872fe431bb6c0afa                       1321   
..                                          ...                        ...   
295  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd                       1287   
296  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd                       1288   
297  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd                       1289   
298  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd                       1290   
299  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd                       1291   

     transcoders_pools_rewardTokens  
0                         70.798464  
1                         71.009434  
2                         71.315195  
3                         71.683247  
4                         71.929907  
..                              ...  
295                      247.761644  
296                      249.825870  
297                      259.301906  
298                      259.894689  
299                      261.312646  

[300 rows x 3 columns]

In [52]:
pool_rewards[:35]

transcoders_id transcoders_pools_round_id  \
0   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1317   
1   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1318   
2   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1319   
3   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1320   
4   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1321   
5   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1322   
6   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1323   
7   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1324   
8   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1325   
9   0x9c10672cee058fd658103d90872fe431bb6c0afa                       1326   
10  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1327   
11  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1328   
12  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1329   
13  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1330   
14  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1331   
15  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1333   
16  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1338   
17  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1355   
18  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1356   
19  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1357   
20  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1358   
21  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1359   
22  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1360   
23  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1361   
24  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1362   
25  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1363   
26  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1365   
27  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1367   
28  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1368   
29  0x9c10672cee058fd658103d90872fe431bb6c0afa                       1369   
30  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9                       1081   
31  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9                       1082   
32  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9                       1083   
33  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9                       1087   
34  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9                       1088   

    transcoders_pools_rewardTokens  
0                        70.798464  
1                        71.009434  
2                        71.315195  
3                        71.683247  
4                        71.929907  
5                        72.130936  
6                        72.177503  
7                        72.689197  
8                        72.922845  
9                        73.031857  
10                       79.594061  
11                       79.867596  
12                       80.110208  
13                       80.385053  
14                      235.099478  
15                      236.111295  
16                      235.624113  
17                      162.967684  
18                      163.431744  
19                      164.334113  
20                      164.116658  
21                      164.635347  
22                      165.155301  
23                      165.765275  
24                      166.288093  
25                      166.883702  
26                      169.068481  
27                      169.189251  
28                      170.266762  
29                      170.509361  
30                       45.776365  
31                       66.914339  
32                       66.941976  
3

In [113]:
grouped = pool_rewards.groupby('transcoders_id')
rmin = grouped.min()['transcoders_pools_round_id'].rename('min').astype('int64')
rmax = grouped.max()['transcoders_pools_round_id'].rename('max').astype('int64')

In [114]:
top_transcoders = pd.concat([rmin,rmax],axis=1).reset_index()

In [121]:
# i'd much rather count null values than go thru this, but they're working on that bug
top_transcoders['diff'] = top_transcoders['max'] - top_transcoders['min']
# inclusive so 29 = 30 rounds
top_transcoders['missing'] = top_transcoders['diff'] - 29
top_transcoders['implied_rate'] = 1 - (top_transcoders['missing'] / 30)
top_transcoders

transcoders_id   min   max  diff  missing  \
0  0x4f4758f7167b18e1f5b3c1a7575e3eb584894dbc   962   991    29        0   
1  0x4ff088ac5422f994486663ff903b040692797168  1068  1101    33        4   
2  0x525419ff5707190389bfb5c87c375d710f5fcb0e  1263  1292    29        0   
3  0x731808ad8b1c3d13e8972db838ada5fc6ae3c2c8  2412  2442    30        1   
4  0x942f0c28fb85ea0b50bfb76a3ecfa99861fa9b4b  1787  1816    29        0   
5  0x9c10672cee058fd658103d90872fe431bb6c0afa  1317  1369    52       23   
6  0xd84781e1a9b74d71ea76cda8bb9f30893bfd00d1  1284  1313    29        0   
7  0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9  1081  1113    32        3   
8  0xdac817294c0c87ca4fa1895ef4b972eade99f2fd  1262  1291    29        0   
9  0xe0a4a877cd0a07da7c08dffebc2546a4713147f2  1413  1442    29        0   

   implied_rate  
0      1.000000  
1      0.866667  
2      1.000000  
3      0.966667  
4      1.000000  
5      0.233333  
6      1.000000  
7      0.900000  
8      1.000000  
9      1.000000

In [130]:
num_transcoders = lpt.Query.transcoders(
    first=500
)

sg.query_df([
    num_transcoders.id
])

transcoders_id
0    0x0000000000000000000000000000000000000000
1    0x004dcf13e2f37abea365a9ed94e568906a39a5c7
2    0x00803b76dc924ceabf4380a6f9edc2ddd3c90f38
3    0x00ac9c5193660b1f69092fb4b0f011521b67627f
4    0x00b1606fc5b771f3079b4fd3ea49e66a2d5fd665
..                                          ...
343  0xfd9803ced16490c7df31c258d9d0535ebabd654c
344  0xfdcd7840908990db1915b2e9d1ae68af6086a00b
345  0xff8c740244183195bb6ec172c9686801266b0f99
346  0xffa340bc544de73148d2c4c39db8b7e51f5d4adc
347  0xffdfa897dd098bfda8af8cf3fb76c34cc2d3bfb2

[348 rows x 1 columns]

So, there are 348 total transcoders in the system. Looks like there is a system "limit 100" 

In [131]:
sg.query_df([
    lpt.Query.transcoders.id
])

transcoders_id
0   0x0000000000000000000000000000000000000000
1   0x004dcf13e2f37abea365a9ed94e568906a39a5c7
2   0x00803b76dc924ceabf4380a6f9edc2ddd3c90f38
3   0x00ac9c5193660b1f69092fb4b0f011521b67627f
4   0x00b1606fc5b771f3079b4fd3ea49e66a2d5fd665
..                                         ...
95  0x43892448e91d987ebf7b764d0af6675499a1a4ba
96  0x44829242e3301b16f8823093affd3e9f8e71c36c
97  0x44c98af5da6a31c422ee5eed18d53986dd9bc01e
98  0x46decd8532b7a794e4e786a992ee672e1fec733a
99  0x4712e01e944802613de3a0a6d23274e7e0243015

[100 rows x 1 columns]

In [5]:
active_transcoders = transcoders(
    orderBy=transcoders.totalStake,
    orderDirection='desc',
    first=1000,
    where=[
        transcoders.active == True
    ]
)

sg.query_df([
    active_transcoders.id
])

transcoders_id
0   0x9c10672cee058fd658103d90872fe431bb6c0afa
1   0xda43d85b8d419a9c51bbf0089c9bd5169c23f2f9
2   0x525419ff5707190389bfb5c87c375d710f5fcb0e
3   0x4ff088ac5422f994486663ff903b040692797168
4   0x942f0c28fb85ea0b50bfb76a3ecfa99861fa9b4b
..                                         ...
95  0x44c98af5da6a31c422ee5eed18d53986dd9bc01e
96  0xa408138c3c75c927b8614a24a39de02ffd71405f
97  0xb3ec176da90524ae3ea530caca4d8afc84f10c09
98  0x3a025488a37fecdb6ca8a4cbd72ba87a919d8bf3
99  0x5b5079b69a3f8dc07ca05ee301e39a08a002def3

[100 rows x 1 columns]

In [6]:
inactive_transcoders = transcoders(
    orderBy=transcoders.totalStake,
    orderDirection='desc',
    first=1000,
    where=[
        transcoders.active == False
    ]
)

sg.query_df([
    inactive_transcoders.id
])

transcoders_id
0    0x481efb669b423cfaffa49890205427dea9b3d693
1    0xbe9a8b21353ea73794553c154c6c9a06823fcd11
2    0x50d69f8253685999b4c74a67ccb3d240e2a56ed6
3    0xe1c879f065682a0d35966596a72d5681a56118a5
4    0x345551571c5ef20111c6168b9a498dfb836e7c09
..                                          ...
243  0x04c670b89907d9740c2f8683e397096d040b0682
244  0x04008b69b7af1f94c1233423d2539d517fc30a74
245  0x00ac9c5193660b1f69092fb4b0f011521b67627f
246  0x004dcf13e2f37abea365a9ed94e568906a39a5c7
247  0x0000000000000000000000000000000000000000

[248 rows x 1 columns]